service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/UploadAdditionalApplicationsController.java

```diff
@@ -0,0 +1,179 @@
+@RequiredArgsConstructor(onConstructor = @__(@Autowired))
+public class UploadAdditionalApplicationsController extends CallbackController {
+    private static final String DISPLAY_AMOUNT_TO_PAY = "displayAmountToPay";
+    private static final String AMOUNT_TO_PAY = "amountToPay";
+    private static final String TEMPORARY_C2_DOCUMENT = "temporaryC2Document";
+    private static final String TEMPORARY_OTHER_APPLICATIONS_BUNDLE = "temporaryOtherApplicationsBundle";
+    private final PaymentService paymentService;
+    private final PbaNumberService pbaNumberService;
+    private final UploadAdditionalApplicationsService uploadC2DocumentsService;
+    private final ApplicationsFeeCalculator applicationsFeeCalculator;
+
+    @PostMapping("/get-fee/mid-event")
+    public AboutToStartOrSubmitCallbackResponse handleMidEvent(@RequestBody CallbackRequest callbackRequest) {
+        CaseDetails caseDetails = callbackRequest.getCaseDetails();
+        CaseData caseData = getCaseData(caseDetails);
+
+        caseDetails.getData().putAll(applicationsFeeCalculator.calculateFee(caseData));
+
+        return respond(caseDetails);
+    }
+
+    @PostMapping("/validate/mid-event")
+    public AboutToStartOrSubmitCallbackResponse handleValidateMidEvent(@RequestBody CallbackRequest callbackRequest) {
+        CaseDetails caseDetails = callbackRequest.getCaseDetails();
+        CaseData caseData = getCaseData(caseDetails);
+
+        List<String> errors = new ArrayList<>();
+
+        if (caseData.getTemporaryPbaPayment() != null) {
+            var updatedPbaNumber = pbaNumberService.update(caseData.getTemporaryPbaPayment().getPbaNumber());
+            caseDetails.getData().put("temporaryPbaPayment",
+                caseData.getTemporaryPbaPayment().toBuilder().pbaNumber(updatedPbaNumber).build());
+            errors.addAll(pbaNumberService.validate(updatedPbaNumber));
+        }
+
+        return respond(caseDetails, errors);
+    }
+
+    @PostMapping("/about-to-submit")
+    public AboutToStartOrSubmitCallbackResponse handleAboutToSubmit(
+        @RequestBody CallbackRequest callbackRequest) {
+        CaseDetails caseDetails = callbackRequest.getCaseDetails();
+        CaseData caseData = getCaseData(caseDetails);
+
+        List<Element<AdditionalApplicationsBundle>> additionalApplications = defaultIfNull(
+            caseData.getAdditionalApplicationsBundle(), new ArrayList<>()
+        );
+
+        C2DocumentBundle c2DocumentBundle = null;
+        OtherApplicationsBundle otherApplicationsBundle = null;
+
+        if (hasC2Order(caseData)) {
+            c2DocumentBundle = uploadC2DocumentsService.buildC2DocumentBundle(caseData);
+        }
+
+        if (hasOtherApplicationsOrder(caseData)) {
+            otherApplicationsBundle = uploadC2DocumentsService.buildOtherApplicationsBundle(caseData);
+        }
+
+        additionalApplications.add(0, element(uploadC2DocumentsService.buildAdditionalApplicationsBundle(
```

Comment:
uploadC2DocumentsService.buildAdditionalApplicationsBundle could accept just case data and do internally check if there is c2DocumentBundle or otherApplicationsBundle

Reply:
updated